src/cli/commands/add.js

```diff
@@ -89,7 +97,11 @@ export class Add extends Install {
    */
 
   async savePackages(): Promise<void> {
-    const {dev, exact, tilde, optional, peer} = this.flags;
+    const {exact, tilde} = this.flags;
+
+    // fill rootPatternsToOrigin without `excludePatterns`
+    await Install.prototype.fetchRequestFromCwd.call(this);
```

Comment:
Tried to explain it in the comment, but `this.fetchRequestFromCwd` passes the arguments along, so when it gets to [cli/commands/install.js#L197](https://github.com/yarnpkg/yarn/blob/master/src/cli/commands/install.js#L197) the packages in the arguments never end up with the map of `{ package: dependencyType }` which is what we want. They would just be excluded.